flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java

```diff
@@ -0,0 +1,370 @@
+
+		Set<TopicPartition> newPartitions = Collections.unmodifiableSet(partitionChange.getNewPartitions());
+		OffsetsInitializer.PartitionOffsetsRetriever offsetsRetriever = getOffsetsRetriever();
+
+		Map<TopicPartition, Long> startingOffsets =
+				startingOffsetInitializer.getPartitionOffsets(newPartitions, offsetsRetriever);
+		Map<TopicPartition, Long> stoppingOffsets =
+				stoppingOffsetInitializer.getPartitionOffsets(newPartitions, offsetsRetriever);
+
+		Set<KafkaPartitionSplit> partitionSplits = new HashSet<>(newPartitions.size());
+		for (TopicPartition tp : newPartitions) {
+			Long startingOffset = startingOffsets.get(tp);
+			long stoppingOffset = stoppingOffsets.getOrDefault(tp, KafkaPartitionSplit.NO_STOPPING_OFFSET);
+			partitionSplits.add(new KafkaPartitionSplit(tp, startingOffset, stoppingOffset));
+		}
+		discoveredPartitions.addAll(newPartitions);
+		return new PartitionSplitChange(partitionSplits, partitionChange.getRemovedPartitions());
+	}
+
+	// This method should only be invoked in the coordinator executor thread.
+	private void handlePartitionSplitChanges(PartitionSplitChange partitionSplitChange, Throwable t) {
+		if (t != null) {
+			throw new FlinkRuntimeException("Failed to handle partition splits change due to ", t);
+		}
+		if (partitionDiscoveryIntervalMs < 0) {
+			LOG.debug("");
+			noMoreNewPartitionSplits = true;
+		}
+		// TODO: Handle removed partitions.
+		addPartitionSplitChangeToPendingAssignments(partitionSplitChange.newPartitionSplits);
+		assignPendingPartitionSplits();
+	}
+
+	// This method should only be invoked in the coordinator executor thread.
+	private void addPartitionSplitChangeToPendingAssignments(Collection<KafkaPartitionSplit> newPartitionSplits) {
+		int numReaders = context.currentParallelism();
+		for (KafkaPartitionSplit split : newPartitionSplits) {
+			// TODO: Implement a more sophisticated algorithm to reduce partition movement when parallelism changes.
+			int ownerReader = split.getTopicPartition().hashCode() % numReaders;
+			pendingPartitionSplitAssignment.computeIfAbsent(ownerReader, r -> new HashSet<>()).add(split);
+		}
+		LOG.debug("Assigned {} to {} readers of consumer group {}.", newPartitionSplits, numReaders, consumerGroupId);
+	}
+
+	// This method should only be invoked in the coordinator executor thread.
+	private void assignPendingPartitionSplits() {
+		Map<Integer, List<KafkaPartitionSplit>> incrementalAssignment = new HashMap<>();
+		pendingPartitionSplitAssignment.forEach((ownerReader, pendingSplits) -> {
+			if (!pendingSplits.isEmpty() && context.registeredReaders().containsKey(ownerReader)) {
+				// The owner reader is ready, assign the split to the owner reader.
+				incrementalAssignment.computeIfAbsent(ownerReader, r -> new ArrayList<>()).addAll(pendingSplits);
+			}
+		});
+		if (incrementalAssignment.isEmpty()) {
+			// No assignment is made.
+			return;
+		}
+		context.assignSplits(new SplitsAssignment<>(incrementalAssignment));
```

Comment:
nit: can you add a INFO log after this? it can be very useful if we need to debug assignment problem.